### PR TITLE
Backend; admin approvals API

### LINKED ISSUE #214: [Backend] - Admin/Approvals API
As a developer, I would like an API that validates and changes the approval status of a submitted dataset and/or edits a submitted dataset.

**Description** 

An admin has the capability to approve/reject datasets as well as add comments. This is a feature ONLY for admins. Therefore, there needs to be a way to check whether the user is an admin, and react accordingly (such as giving them access to the page to approve/reject and modify datasets). This API should change the approved status of a data set and enter relevant comments about the dataset into its DB row as well as let the admin modify the dataset selected.

Additional Notes:
- If an admin rejects a dataset then it must also be deleted from the DB
- When an admin raises a flag, a message X must be sent to the uploader indicating that their dataset needs editing with message X

2 types of comments need to be saved. 
1 ) Comment describing the dataset - (Kiyanda: I love this dataset. Its my favourite dataset). This comment is appended to the existing dataset comment (found in the actual dataset)
2) Comment regarding the status of the submitted dataset (incorrect entry) 

**Acceptance Criteria**
- Allow only admins to access the approval/reject page
- Allow admins to modify/edit datasets
- Allow admin to approve/reject the dataset
- Allow admin to comment on a dataset
- Allow admin to flag an incorrect submission

Verify if multiple apis need to be created to support these different commands

### PR DESCRIPTION
**Pull request by silent-po2**
This PR contains all the backend functionality needed for admin approvals; getting, approving, flagging, and rejecting unapproved data sets, editing of data sets, and also setting all newly uploaded data sets as unapproved.

### CODE REVIEW

File: Server/src/controllers/DataSetController.ts [hunk 1]
```diff
@@ -0,0 +1,245 @@
+import { Request, Response } from 'express';
+import { DataSetService } from '../services/DataSetService';
+import { IDataRequestModel } from "../models/interfaces/DataRequestModelInterface";
+
+export class DataSetController {
+    private processedRequest: IDataRequestModel;
+    private dataSetService: DataSetService
+
+    constructor() {
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts:9]: is it a requirement to add an empty constructor?

File: Server/src/controllers/DataSetController.ts [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+        let validateData = this.validateInputData(request)
+        if (!validateData) {
+            response.status(400).json("Invalid search params entered");
+        }
+        else {
+            this.dataSetService = new DataSetService();
+            let requestParams: any = { ...request.query };
+            this.processedRequest = requestParams;
+            try {
+                let arrayOfData = await this.dataSetService.getArrayOfDatasets(this.processedRequest)
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts:33]: imo getDatasets is a better name, the plural implies an array + we can check the returning type, right?

File: Server/src/controllers/DataSetController.ts [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+     * send it to the getDataService to acquire an array of data sets that match the entered search terms, and
+     * then return a response with the array of data sets. If the request is invalid, or some other problem arrises, 
+     * it will throw an error.
+     * 
+     * @param request
+     * An object containing the request information and parameters: Request 
+     * @param response 
+     * An object containing a response: Response
+     */
+    async createRequestForData(request: Request, response: Response) {
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts:23]: very ambigous method name, unless it does something generic, but it doesnt, maybe call it createRequestForDataQuerying or requestDataByQuery

File: Server/src/controllers/DataSetController.ts [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+        if (!validateData) {
+            response.status(400).json("Invalid search params entered");
+        }
+        else {
+            this.dataSetService = new DataSetService();
+            let requestParams: any = { ...request.query };
+            this.processedRequest = requestParams;
+            try {
+                let arrayOfData = await this.dataSetService.getArrayOfDatasets(this.processedRequest)
+                return response.status(200).send(arrayOfData);
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts]: shouldnt we use response.json(arrayOfData)?

File: Server/src/controllers/DataSetController.ts [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+            } catch (err) {
+                response.status(400).send(err);
+            }
+        }
+    }
+
+    async createRequestToDeleteSavedDataSet(request: Request, response: Response) {
+        let requestParam = request.params.datasetId;
+        let datasetId = Number(requestParam);
+        if (isNaN(datasetId)) {
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts:44]: just keep in mind that backend expects ids as numbers when we want to move to UUID, on a side note, what would we gain from converting the ids to UUID? maybe we should leave it as is

File: Server/src/controllers/DataSetController.ts [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+            response.status(400).json("Invalid data set ID entered");
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts:45]: here an error is returned as a json, but in the method above it is returned with response.send() method, what is the difference?

File: Server/src/controllers/DataSetController.ts [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+        }
+        else {
+            let userId: number = request.body.user.account_id
+            try {
+                let requestResponse = await this.dataSetService.removeSavedDatasetService(userId, datasetId)
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts]: imo saved is redundant, since it is being removed from DB, obviously it is saved, also, the method name says removeSavedDatasetService, does it remove a dataset or does it remove a service?

File: Server/src/controllers/DataSetController.ts [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+        let userId: any = request.body.user.account_id
+        if (isNaN(userId)) {
+            response.status(400).send("Invalid search params entered");
+        }
+        else {
+            try {
+                let arrayOfData = await this.dataSetService.getUserUploadedDatasets(userId)
+                return response.status(200).json(arrayOfData);
+            } catch (err) {
+                response.status(500).send(err);
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts]: dont errors contain status code? because we are raising new InternalServerError()? why do we hardcode the exception status code?

File: Server/src/controllers/DataSetController.ts [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+     * An object containing a response: Response
+     */
+    async createRequestForUserSavedDatsets(request: Request, response: Response) {
+        let userId: any = request.body.user.account_id
+        if (isNaN(userId)) {
+            response.status(500).json("Invalid search params entered");
+        }
+        else {
+            try {
+                let arrayOfData = await this.dataSetService.getUserSavedDatasets(userId)
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts]: what is the difference between user saved and user uploaded datsaets?

File: Server/src/controllers/DataSetController.ts [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+                return response.status(200).send(arrayOfData);
+            } catch (err) {
+                response.status(500).send(err);
+            }
+        }
+    }
+
+    //TODO: Readd later
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts:103]: Readd later todo?

File: Server/src/controllers/DataSetController.ts [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+     * containing the unapproved data sets in the database
+     * 
+     * @param request
+     * An object containing the request information and parameters: Request 
+     * @param response 
+     * An object containing a response: Response
+     */
+    async createRequestForUnapprovedDatsets(request: Request, response: Response) {
+        try {
+            this.dataSetService = new DataSetService();
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts:142]: this is what i mean, here a new instance is created, but in other cr3eateRequests method its not, are they guaranteed to have datsetService instantiated?

File: Server/src/controllers/DataSetController.ts [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+
+    /**
+     * This method verifies that a request has at least one valid search parameter and returns true if so
+     * and false if not.
+     * 
+     * @param request 
+     * An object containing the request information and parameters: Request
+     */
+    private validateInputData(request: Request) {
+        if (request.query.hasOwnProperty('datasetId') || request.query.hasOwnProperty('material')
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts:158]: i think in the future we could replace these checks with Yup

File: Server/jest.config.js [hunk 13]
```diff
@@ -13,12 +13,12 @@ module.exports = {
         ".mock.ts"
     ],
     collectCoverageFrom: [
-    'src/controllers/*.ts',
-    'src/middleware/*.ts',
-    'src/models/*.ts',
-    'src/services/*.ts'
-  ]
-	,
+        'src/controllers/**/*.ts',
```
Comment (TiraFire) [inline Server/jest.config.js:16]: is this part of the admin story or is this the fix you did for JT?

File: Server/src/controllers/DataSetController.ts [hunk 14]
```diff
@@ -0,0 +1,245 @@
...
+    async createRequestToFlagDataset(request: Request, response: Response) {
+        if (!request.query && !request.query.datasetId) {
+            response.status(400).json("No datasetID provided to flag dataset");
+        }
+        else {
+            let datasetIdToFlag = Number(request.query.datasetId)
+            let flaggedComments = String(request.query.flaggedComments)
+            let additionalComments = String(request.query.additionalComments)
+            try {
+                let requestResponse = await this.dataSetService.flagNewDataset(datasetIdToFlag, flaggedComments, additionalComments)
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts:176]: does the method check that it is the admin who flags and not an ordinary user?

File: Server/src/controllers/DataSetController.ts [hunk 15]
```diff
@@ -0,0 +1,245 @@
...
+     * An object containing a response: Response
+     */
+    async createRequestForUserUploadedDatasets(request: Request, response: Response) {
+        let userId: any = request.body.user.account_id
+        if (isNaN(userId)) {
+            response.status(400).send("Invalid search params entered");
+        }
+        else {
+            try {
+                let arrayOfData = await this.dataSetService.getUserUploadedDatasets(userId)
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts]: this.dataSetService won't this be null? i dont see the service being created in the constructor

File: Server/src/controllers/DataSetController.ts [hunk 16]
```diff
@@ -0,0 +1,245 @@
...
+            try {
+                let requestResponse = await this.dataSetService.flagNewDataset(datasetIdToFlag, flaggedComments, additionalComments)
+                return response.status(requestResponse.statusCode).json(requestResponse.message);
+            } catch (error) {
+                response.status(error.status).json(error.message);
+            }
+        }
+    }
+
+    async createRequestToRejectDataset(request: Request, response: Response) {
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts:184]: the file is quite big, I know we are rushing now a little, but next sprint lets reuse validation checks and try catch it should be pretty easy with JS, i have a few ideas ping me if interested

File: Server/src/controllers/DataSetController.ts [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+        if (isNaN(datasetId)) {
+            response.status(400).json("Invalid data set ID entered");
+        }
+        else {
+            try {
+                this.dataSetService = new DataSetService();
+                let requestResponse = await this.dataSetService.rejectDataSet(datasetId)
+                return response.status(requestResponse.statusCode).json(requestResponse.message);
+            } catch (error) {
+                response.status(error.status).json(error.message);
```
Comment (ds161833) [inline Server/src/controllers/DataSetController.ts:196]: so here we are replying with errors.status and json(error.message) this looks good, then the above try catch should be the same, might as well just put the try catch in a separate funciton and wrap the core code in it like so:
withTryCatch(() => yourcode())
const withTryCatch(delegate) => {
  try {
    delegate()
  }
  catch(error)  {
    your typical handler
  }
}

File: Server/src/models/DataUploadModel.ts [hunk 18]
```diff
@@ -40,7 +41,7 @@ export class DataUploadModel {
      * @param publicationTypeReceived 
      * Publication type: string
      */
-    async insertReferenceType(publicationTypeReceived: string): Promise<number> {
+    async insertPreferenceType(publicationTypeReceived: string): Promise<number> {
```
Comment (TiraFire) [inline Server/src/models/DataUploadModel.ts]: Preference type?

File: Server/src/controllers/dataUploadController.ts [hunk 19]
```diff
@@ -9,20 +11,42 @@ import { DataUploadService } from '../services/DataUploadService'
...
             response.status(400).json({
-                message: "No Json to Upload Received"
+                message: "No Data to Upload"
             })
         }
         else {
-            this.dataService = new DataUploadService(request.body);
             try {
+                this.dataSet = request.body
+                this.dataService = new UnapprovedUploadService(this.dataSet)
```
Comment (ds161833) [inline Server/src/controllers/dataUploadController.ts]: dataService shouldnt be any but some interface that UnapprovedUploadService implements and other services that are used in the dataUploadController

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+
+    async selectAllFlaggedDatasets(): Promise<IDatasetIDModel[]> {
+        let allFlaggedDatsets = await this.connection.createQueryBuilder(Unapproveddatasets, 'unapproved_datasets')
+            .select('unapproved_datasets.datasetId', 'dataset_id')
+            .where('unapproved_datasets.isFlagged = 1')
+            .getRawMany();
+        return allFlaggedDatsets
+    }
+
+    async updateApprovedStatus(datasetId: number) {
```
Comment (TiraFire) [inline Server/src/models/DatasetModels/DatasetApprovalModel.ts]: a bit vague, update sounds like can be either set to 1 or 0 but this only approves. can we change it to make sure people know this approves a dataset?

File: Server/src/controllers/DataSetController.ts (hunk 1 above)
Comment (GD-Dheer) [reply Server/src/controllers/DataSetController.ts:9]: Yeah. Or else we must have static methods

File: Server/src/controllers/DataSetController.ts (hunk 3 above)
Comment (GD-Dheer) [reply Server/src/controllers/DataSetController.ts:23]: Will change. This was one of the first controllers we created. We did not completely change the old code in the pre-existing dataset controller

File: Server/src/controllers/DataSetController.ts (hunk 4 above)
Comment (GD-Dheer) [reply Server/src/controllers/DataSetController.ts]: Yeah! .json is what we are using in most of the codebase. Will update

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+        await this.updateDatasetComments(datasetId, additionalComment)
+        await this.connection.createQueryBuilder(Unapproveddatasets, 'unapproved_datasets')
+            .update('unapproved_datasets')
+            .set({ flaggedComments: flaggedComment, isFlagged: 1 })
+            .where('unapproved_datasets.datasetId = :datasetId', { datasetId: datasetId })
+            .execute()
+        return "Dataset Flagged!"
+    }
+
+    async approveDataset(datasetId: number) {
```
Comment (TiraFire) [inline Server/src/models/DatasetModels/DatasetApprovalModel.ts:66]: is this the only method that uses the updateApprovedStatus method?

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts [hunk 22]
```diff
@@ -0,0 +1,88 @@
...
+
+export class DatasetApprovalModel {
+    private connection: Connection;
+    private commonModel: DatasetCommonModel
+    constructor() {
+        this.connection = getConnection();
+        this.commonModel = new DatasetCommonModel();
+    }
+
+    private selectDatasetCommentQuery = (id: number) =>
```
Comment (ds161833) [inline Server/src/models/DatasetModels/DatasetApprovalModel.ts:17]: shouldnt this be async?

Comment (ds161833) [reply Server/src/models/DatasetModels/DatasetApprovalModel.ts:17]: also does it have return type?

File: Server/src/controllers/DataSetController.ts (hunk 5 above)
Comment (GD-Dheer) [reply Server/src/controllers/DataSetController.ts:44]: Auto-increment ID is similar to using a randomly generated unique UUID. I suppose the main benefit would be management if theres ever many many users. Complex systems adopt UUIDs. then again ours is not so complex, we can reconsider the change of PK

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+        this.connection.createQueryBuilder(Dataset, 'dataset')
+            .select('dataset.comments', 'comments')
+            .where('dataset.id = :id', { id: id })
+            .getRawOne();
+
+    async selectUserFlaggedDatasets(uploaderId: number): Promise<IDatasetIDModel[]> {
+        let userFlaggedDatasets = await this.connection.createQueryBuilder(Unapproveddatasets, 'unapproved_Datasets')
```
Comment (ds161833) [inline Server/src/models/DatasetModels/DatasetApprovalModel.ts]: it is unapproved_Datasets in createQueryBuilder but unapproved_datasets in the select, is it fine?

File: Server/src/controllers/DataSetController.ts (hunk 7 above)
Comment (GD-Dheer) [reply Server/src/controllers/DataSetController.ts]: Yeah. Good catch on some of the naming.... Will change

File: Server/src/controllers/DataSetController.ts (hunk 8 above)
Comment (GD-Dheer) [reply Server/src/controllers/DataSetController.ts]: Old

File: Server/src/controllers/DataSetController.ts (hunk 9 above)
Comment (GD-Dheer) [reply Server/src/controllers/DataSetController.ts]: When a user uploaded to the database, they are counted as user uploaded dataset, a way to identify the uploader

Saved is just their favorite datasets

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts [hunk 24]
```diff
@@ -0,0 +1,88 @@
...
+
+    async approveDataset(datasetId: number) {
+        await this.updateApprovedStatus(datasetId)
+        await this.commonModel.wipeEntryFromUnapprovedTable(datasetId)
+        return "Successfully approved new Dataset"
+    }
+
+    async updateDatasetComments(datasetId: number, datasetCommentsToAppend?: string) {
+        let oldComment = await this.selectDatasetCommentQuery(datasetId)
+        let newComment = oldComment.comments.concat(" " + datasetCommentsToAppend)
```
Comment (ds161833) [inline Server/src/models/DatasetModels/DatasetApprovalModel.ts:74]: so comments are getting appended? shouldnt we just override the previous comment? why is it appended

File: Server/src/controllers/DataSetController.ts (hunk 12 above)
Comment (GD-Dheer) [reply Server/src/controllers/DataSetController.ts:158]: Already planned

File: Server/src/models/DatasetModels/DatasetDeleteModel.ts [hunk 25]
```diff
@@ -0,0 +1,276 @@
...
+    }
+
+    async deletePublicationType(publicationTypeId: number) {
+        let isPublicationTypeInUse = await this.selectOneUseOfPublicationTypeQuery(publicationTypeId)
+        if (isPublicationTypeInUse == undefined) {
+            await this.deletePublicationTypeQuery(publicationTypeId)
+        }
+    }
+
+    async rejectDataset(id: number): Promise<string> {
```
Comment (TiraFire) [inline Server/src/models/DatasetModels/DatasetDeleteModel.ts:258]: hopefully this wont cause issues, seems like it could be a source of bugs if not properly done, will need some testing for sure to assure this deletes a dataset properly without remains or without ruining other datasets, but im sure you guys did it well :P

File: Server/src/controllers/DataSetController.ts (hunk 11 above)
Comment (GD-Dheer) [reply Server/src/controllers/DataSetController.ts:142]: It is missing from the other methods.... Was forgotten to add. Good catch

File: Server/jest.config.js (hunk 13 above)
Comment (GD-Dheer) [reply Server/jest.config.js:16]: Independent of the admins

File: Server/src/controllers/DataSetController.ts (hunk 14 above)
Comment (GD-Dheer) [reply Server/src/controllers/DataSetController.ts:176]: JWT middleware handles this

File: Server/src/models/DatasetModels/DatasetDeleteModel.ts [hunk 26]
```diff
@@ -0,0 +1,276 @@
...
+        // Delete dataset proper
+        await this.commonModel.wipeEntryFromUnapprovedTable(id)
+        await this.deleteDatasetQuery(id)
+        await this.deleteDatasetDataType(rawDatasetFKs.datatypeId)
+        // Delete publication and its data
+        await this.deleteAuthorsOfPublication(rawDatasetFKs.publicationId)
+        await this.deletePublication(rawDatasetFKs.publicationId)
+        await this.deletePublisher(rawDatasetFKs.publisherId)
+        await this.deletePublicationType(rawDatasetFKs.publicationTypeId)
+        return "Successfully deleted data set"
```
Comment (TiraFire) [inline Server/src/models/DatasetModels/DatasetDeleteModel.ts]: maybe change to "Successfully removed data set."? deleted seems... harsh

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts (hunk 24 above)
Comment (ds161833) [reply Server/src/models/DatasetModels/DatasetApprovalModel.ts:74]: did you test that it works the way you want it to? because concat accepts array as an argument https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/concat
and a string is an array of chars, if you wanna append one to an array use .push(string) instead

File: Server/src/controllers/DataSetController.ts (hunk 15 above)
Comment (GD-Dheer) [reply Server/src/controllers/DataSetController.ts]: Similar like above, it seems like we forgot to instantiate the service in some of the older methods

File: Server/src/controllers/DataSetController.ts (hunk 17 above)
Comment (GD-Dheer) [reply Server/src/controllers/DataSetController.ts:196]: Good idea

File: Server/src/models/DatasetModels/DatasetDeleteModel.ts [hunk 27]
```diff
@@ -0,0 +1,276 @@
...
+import { Representations } from "../entities/Representations";
+import { Unapproveddatasets } from "../entities/Unapproveddatasets";
+import { Units } from "../entities/Units";
+import { DatasetCommonModel } from "./DatasetCommonModel";
+
+
+/**
+ * This model class is responsible for updating the database with the extracted from the fileUpload. 
+ */
+export class DatasetDeleteModel {
```
Comment (ds161833) [inline Server/src/models/DatasetModels/DatasetDeleteModel.ts:20]: is this the best name for the model that is responsible for updating db with extracted dataset?

File: Server/src/routes/DatasetRouter.ts [hunk 28]
```diff
@@ -0,0 +1,52 @@
...
+
+router.get('/api/v1/dataset*', (request: Request, response: Response) => {
+    dataSetController.createRequestForData(request, response);
+});
+
+router.delete('/api/v1/dataset/:datasetId', [JWTAuthenticator.verifyJWT, JWTAuthenticator.verifyAdmin], (request: Request, response: Response) => {
+    dataSetController.createRequestToRejectDataset(request, response)
+})
+
+router.put('/api/v1/flagDataSet/:datasetId', [JWTAuthenticator.verifyJWT, JWTAuthenticator.verifyAdmin], (request: Request, response: Response) => {
```
Comment (ds161833) [inline Server/src/routes/DatasetRouter.ts:36]: a more conventional route would be api/v1/dataset/:datsetId/flag and it owuld be a patch method, because UPDATE 
 is meant to update entity completely (all fields are updated so you must pass a new dataset in the request.body), patch updates entity partially (one field for example)

**Review by ds161833 — COMMENTED** [review]
good job, resolve the comments and let me know

File: Server/src/services/DataUpload/EditUploadService.ts [hunk 29]
```diff
@@ -0,0 +1,95 @@
...
+import { DatasetCommonModel } from '../../models/DatasetModels/DatasetCommonModel';
+import { DatasetDeleteModel } from '../../models/DatasetModels/DatasetDeleteModel';
+import { DataUploadModel } from "../../models/DataUploadModel"
+import { Authors } from "../../models/entities/Authors"
+import AbstractUploadService from "./AbstractUploadService"
+
+
+export default class EditUploadService extends AbstractUploadService {
+    async uploadData(): Promise<IResponse> {
+        let deleteModel = new DatasetDeleteModel()
```
Comment (TiraFire) [inline Server/src/services/DataUpload/EditUploadService.ts:13]: is there a better way of updating a dataset than to delete it all and re-add it? seems like this is a lot of work for what it might do

File: Server/src/controllers/DataSetController.ts (hunk 3 above)
Comment (ds161833) [reply Server/src/controllers/DataSetController.ts:23]: ok i see

File: Server/src/tests/services/DataSetService.spec.ts [hunk 30]
```diff
@@ -51,7 +51,7 @@ describe('data service test', () => {
     let testData: IDataRequestModel;
     testData = {} as any;
     testData.datasetId = undefined;
-    testData.material = ["carbon, graphite, pressed graphite"];
+    testData.material = ["O2", "carbon, graphite, pressed graphite"];
```
Comment (TiraFire) [inline Server/src/tests/services/DataSetService.spec.ts:54]: why the change?

File: Server/src/controllers/DataSetController.ts (hunk 9 above)
Comment (ds161833) [reply Server/src/controllers/DataSetController.ts]: oh right, i forgot

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts (hunk 20 above)
Comment (ds161833) [reply Server/src/models/DatasetModels/DatasetApprovalModel.ts]: agreed

File: Server/src/models/DatasetModels/DatasetDeleteModel.ts (hunk 26 above)
Comment (ds161833) [reply Server/src/models/DatasetModels/DatasetDeleteModel.ts]: very minor + this is backend, consumers are literally executing a delete request, the FE will not display the text the backend supplies on success

**Review by TiraFire — COMMENTED** [review]
be careful with some of the class naming. Some start with capitals and some dont, just make sure that if this is intentional that it follows a naming convention

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts (hunk 24 above)
Comment (TiraFire) [reply Server/src/models/DatasetModels/DatasetApprovalModel.ts:74]: The user and kiyandas comments are both relevant, kiyanda just wants the ability to add his own comment on top of the dataset, not override the users

File: Server/src/models/DatasetModels/DatasetDeleteModel.ts (hunk 26 above)
Comment (TiraFire) [reply Server/src/models/DatasetModels/DatasetDeleteModel.ts]: Ahh okay, thought this is what shows when the admin rejects a dataset

File: Server/src/models/DataUploadModel.ts (hunk 18 above)
Comment (GD-Dheer) [reply Server/src/models/DataUploadModel.ts]: I think I meant publicationType lmao

File: Server/src/controllers/dataUploadController.ts (hunk 19 above)
Comment (GD-Dheer) [reply Server/src/controllers/dataUploadController.ts]: Yeah I forgot to get rid of the any

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts (hunk 21 above)
Comment (GD-Dheer) [reply Server/src/models/DatasetModels/DatasetApprovalModel.ts:66]: Yeah

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts (hunk 23 above)
Comment (GD-Dheer) [reply Server/src/models/DatasetModels/DatasetApprovalModel.ts:24]: Will change. This will crash the code

Comment (GD-Dheer) [reply Server/src/models/DatasetModels/DatasetApprovalModel.ts]: Will change. This will crash the code

File: Server/src/models/DatasetModels/DatasetDeleteModel.ts (hunk 26 above)
Comment (GD-Dheer) [reply Server/src/models/DatasetModels/DatasetDeleteModel.ts]: Ideally, I am returning responses to eventually move them into loggers

File: Server/src/models/DatasetModels/DatasetDeleteModel.ts (hunk 27 above)
Comment (GD-Dheer) [reply Server/src/models/DatasetModels/DatasetDeleteModel.ts:20]: Its responsible for holding all db methods regarding deletes

File: Server/src/services/DataUpload/EditUploadService.ts (hunk 29 above)
Comment (GD-Dheer) [reply Server/src/services/DataUpload/EditUploadService.ts:13]: Old values not dont need to be kept if they are not linked to any other datasets

File: Server/src/routes/DatasetRouter.ts (hunk 28 above)
Comment (GD-Dheer) [reply Server/src/routes/DatasetRouter.ts:36]: This route does not contain the dataset. The dataset is only passed on edit or upload.

File: Server/src/tests/services/DataSetService.spec.ts (hunk 30 above)
Comment (GD-Dheer) [reply Server/src/tests/services/DataSetService.spec.ts:54]: Its a test. I think it was just meant to verify

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts (hunk 22 above)
Comment (GD-Dheer) [reply Server/src/models/DatasetModels/DatasetApprovalModel.ts:17]: Will need to move of these methods around since the selects are general methods applicable to the entity and not only the model

File: Server/src/tests/services/DataSetService.spec.ts (hunk 30 above)
Comment (silent-po2) [reply Server/src/tests/services/DataSetService.spec.ts:54]: It originally had O2 and was removed in the search overhaul during some debugging. I realized after the PR was closed that this was not restored and thus this is going back to how it should be, versus a new change.

File: Server/src/services/DataUpload/EditUploadService.ts (hunk 29 above)
Comment (silent-po2) [reply Server/src/services/DataUpload/EditUploadService.ts:13]: Editing a data set will always be a lot of work as you have to edit over a half a dozen database tables. This is the least amount of work required. Going the full edit route is not possible given how some entities, ex. authors, are referenced by multiple tables. You cannot edit an author that is being referenced by a different publication, you have to make a new entry.

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts (hunk 24 above)
Comment (silent-po2) [reply Server/src/models/DatasetModels/DatasetApprovalModel.ts:74]: It is supposed to append as per Kiyanda's requirement

File: Server/src/models/DatasetModels/DatasetDeleteModel.ts (hunk 25 above)
Comment (silent-po2) [reply Server/src/models/DatasetModels/DatasetDeleteModel.ts:258]: It carries out the "order" of deletion in the same manner as how our test migration does. It definitely needs a lot of testing but it falls back on the concept of the data set as the core of everything. 
I wish you were at the group PR review as rejection was explained in depth during it, in part to address these specific concerns.

File: Server/src/models/DatasetModels/DatasetDeleteModel.ts (hunk 27 above)
Comment (silent-po2) [reply Server/src/models/DatasetModels/DatasetDeleteModel.ts:20]: The comment needs some changing as this model is used in updating but it is also used for rejection, which was the initial drive behind these models. I will edit this

**Comment by silent-po2** [comment]
> be careful with some of the class naming. Some start with capitals and some dont, just make sure that if this is intentional that it follows a naming convention

It is the intention of all backend classes to start with a capital eventually. Initially there was not a solid convention, and the PR touches on the first code that the backend ever had, and thus there are gaps that need tweaks.

File: Server/src/models/DatasetModels/DatasetDeleteModel.ts (hunk 26 above)
Comment (silent-po2) [reply Server/src/models/DatasetModels/DatasetDeleteModel.ts]: Changed it to removed

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts (hunk 22 above)
Comment (silent-po2) [reply Server/src/models/DatasetModels/DatasetApprovalModel.ts:17]: Queries do not need to be async and they also do not have a specific return type.

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts (hunk 23 above)
Comment (silent-po2) [reply Server/src/models/DatasetModels/DatasetApprovalModel.ts]: Fixed

File: Server/src/models/DatasetModels/DatasetApprovalModel.ts (hunk 20 above)
Comment (silent-po2) [reply Server/src/models/DatasetModels/DatasetApprovalModel.ts]: Renaming this to: approveSingleDatasetQuery

File: Server/src/controllers/dataUploadController.ts (hunk 19 above)
Comment (silent-po2) [reply Server/src/controllers/dataUploadController.ts]: Fixed in pending commit

File: Server/src/models/DataUploadModel.ts (hunk 18 above)
Comment (silent-po2) [reply Server/src/models/DataUploadModel.ts]: Fixed in pending commit

File: Server/jest.config.js (hunk 13 above)
Comment (silent-po2) [reply Server/jest.config.js:16]: That reminds me to do this for methods, which is required for this story

File: Server/src/controllers/DataSetController.ts (hunk 11 above)
Comment (silent-po2) [reply Server/src/controllers/DataSetController.ts:142]: Added to a number of methods

File: Server/src/controllers/DataSetController.ts (hunk 10 above)
Comment (silent-po2) [reply Server/src/controllers/DataSetController.ts:103]: It was accidentally removed at some point, was thus not updated as changes to this controller were made, and as the frontend plans for these features was dropped, I added this back in as a comment block for next sprint.

File: Server/src/controllers/DataSetController.ts (hunk 15 above)
Comment (silent-po2) [reply Server/src/controllers/DataSetController.ts]: Should be fixed now

File: Server/src/controllers/DataSetController.ts (hunk 7 above)
Comment (silent-po2) [reply Server/src/controllers/DataSetController.ts]: SavedDatasets should all be renamed to UserFavoriteDatasets or similar, once some pending changes are pushed

File: Server/src/controllers/DataSetController.ts (hunk 6 above)
Comment (silent-po2) [reply Server/src/controllers/DataSetController.ts:45]: send was how the file was written initially, the intention is to switch it all to json. Some of these slipped through, thanks for catching them

**Comment by silent-po2** [comment]
All points should be addressed.

**Review by ds161833 — APPROVED** [review]

File: Server/src/tests/services/DataSetService.spec.ts (hunk 30 above)
Comment (TiraFire) [reply Server/src/tests/services/DataSetService.spec.ts:54]: Sounds good

**Review by TiraFire — APPROVED** [review]
